### PR TITLE
Study view refactoring

### PR DESCRIPTION
**Pull request by kalletlak**
This includes
1. Table tooltip for pie chart
2. Delete chart
3. Improve pie chart interaction performance
4. Support multi study

### CODE REVIEW

File: src/pages/studyView/charts/pieChart/PieChart.tsx [hunk 1]
```diff
@@ -139,8 +178,8 @@ class CustomSlice extends React.Component<{}, {}> {
         const d:any = this.props;
         return (
         <g>
-            <Slice {...this.props}/>
-            <title>{`${d.datum.x}:${d.datum.y}`}</title>
+            <Slice {...this.props} active={true}/>
```
Comment (kalletlak) [inline src/pages/studyView/charts/pieChart/PieChart.tsx]: cleanup `active`

File: src/pages/studyView/StudyViewPageStore.ts [hunk 2]
```diff
@@ -146,13 +145,63 @@ export class StudyViewPageStore {
...
+        if(!visibility){
+            //TODO: Currently clears only clinicalDataEqualityFilters,Need to implement for others
+            this._clinicalDataEqualityFilterSet.delete(uniqueKey);
+        }
+        this._chartVisibility.set(uniqueKey, visibility);
+    }
+
+    @action changeChartType(uniqueKey: string, chartType:ChartType, updateDefaultType?:boolean){
+        let chartMetaObject = this._clinicalAttributesMetaSet.get(uniqueKey);
+        if(chartMetaObject){
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts]: is there a reason there would not be a meta object?  just seems strange you don't deal with that case if it in fact is met

File: src/pages/studyView/StudyViewPageStore.ts [hunk 3]
```diff
@@ -146,13 +145,63 @@ export class StudyViewPageStore {
...
+    }
+    @action onChangeSelectedClinicalChart(options: {label: string; value: string;}[]) {
+        this._chartVisibility.clear();
+        for (const option of options) {
+            this.changeChartVisibility(option.value, true)
+        }
+    }
+
+    @computed get clinicalChartOptions() {
+        return _.map(this.clinicalAttributesMeta.result, clinicalAttributeMeta=>({
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts]: if clinicalAttributesMeta is asychronous (i.e. mobxpromise) then the wrapper should also be a moxpromise so that consumer knows they have to check for status=complete.

File: src/pages/studyView/StudyViewPageStore.ts [hunk 4]
```diff
@@ -146,13 +145,63 @@ export class StudyViewPageStore {
...
+            label: clinicalAttributeMeta.clinicalAttribute.displayName,
+            value: clinicalAttributeMeta.uniqueKey
+        }));
+    }
+
+    @computed get selectedClinicalCharts() {
+        return _.reduce(this._chartVisibility.keys(),(acc:{label: string; value: string;}[], next)=>{
+            if(this._chartVisibility.get(next)) {
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts]: i would put === true so it's more clear that you aren't just testing for presence of item in collection, but checking its value.

File: src/pages/studyView/StudyViewPageStore.ts [hunk 5]
```diff
@@ -146,13 +145,63 @@ export class StudyViewPageStore {
...
+                        value: chartMeta.uniqueKey
+                    })
+                }
+            }
+            return acc;
+        },[]);
+    }
+
+    @computed get emptyFilters() {
+        let filters: StudyViewFilter = {} as any;
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts]: just use object literal.  return { studyIds:[this.StudyId]  }

File: src/pages/studyView/StudyViewPageStore.ts [hunk 6]
```diff
@@ -146,13 +145,63 @@ export class StudyViewPageStore {
...
+        filters.studyIds = [this.studyId]
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts]: will we sometimes have more than one studyId?  If so, should we engineer for that?  i.e. this.studyIds    instead of single studyId

File: src/pages/studyView/StudyViewPageStore.ts [hunk 7]
```diff
@@ -283,49 +331,59 @@ export class StudyViewPageStore {
...
+                    .get({ attribute: next.clinicalAttribute, filters: this.emptyFilters })
                     .result!;
                 return acc;
             }, {} as any);
         },
         default: {},
-        onResult:() => {
-            this.initialized=true;
+        onResult:(result) => {
+            if(!this._initialized){
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts]: not sure i understand need for initialization.  we should discuss this on chat.

File: src/pages/studyView/charts/ChartContainer.tsx [hunk 8]
```diff
@@ -90,50 +91,122 @@ export class ChartContainer extends React.Component<IChartContainerProps, {}> {
...
+                this.props.changeChartType(this.props.chartMeta.uniqueKey, ChartType.TABLE, true);
+            }
+        }
+    })
+
+
+
+    @computed get chartWidth(){
+        let chartWidth = styles.chartWidthTwo
+        if(this.props.chartMeta.chartType === ChartType.PIE_CHART){
```
Comment (alisman) [inline src/pages/studyView/charts/ChartContainer.tsx]: how many different chart types are there?

File: src/pages/studyView/charts/ChartContainer.tsx [hunk 9]
```diff
@@ -90,50 +91,122 @@ export class ChartContainer extends React.Component<IChartContainerProps, {}> {
...
+            chartWidth = styles.chartWidthOne
+        }
+        return chartWidth;
     }
 
-    @computed get isChartVisible(){
-        //check if there is chart implementation is present
-        return this.props.chartType === ChartType.PIE_CHART && this.props.clinicalAttribute.datatype==='STRING';
+    @computed get chartHeight(){
+        let chartHeight = styles.chartHeightTwo
```
Comment (alisman) [inline src/pages/studyView/charts/ChartContainer.tsx]: semicolons

File: src/pages/studyView/charts/ChartContainer.tsx [hunk 10]
```diff
@@ -90,50 +91,122 @@ export class ChartContainer extends React.Component<IChartContainerProps, {}> {
...
+        if((this.props.chartMeta.chartType === ChartType.PIE_CHART) || (this.props.chartMeta.chartType === ChartType.TABLE && (this.initialData.result!).length<=3)){
+            chartHeight = styles.chartHeightOne
+        }
+        return chartHeight;
     }
 
-    @computed get data(){
-        return this.props.promise.result || [];
+    @computed get hideLabel(){
+        return this.props.chartMeta.chartType === ChartType.TABLE ? true : false;
```
Comment (alisman) [inline src/pages/studyView/charts/ChartContainer.tsx]: can you just return the condition

File: src/pages/studyView/StudyViewPageStore.ts (hunk 2 above)
Comment (kalletlak) [reply src/pages/studyView/StudyViewPageStore.ts]: `.get` is of type `chartMeta | undefined`. I used `if` condition to check it. I could you post operator`!`

File: src/pages/studyView/charts/ChartContainer.tsx (hunk 8 above)
Comment (kalletlak) [reply src/pages/studyView/charts/ChartContainer.tsx]: there are 5.

Comment (alisman) [reply src/pages/studyView/charts/ChartContainer.tsx]: seems like then are going to have a lot of switch statements and they need to have all five options.  hard to maintain.  would it be better to have subclasses which implement these members?

File: src/pages/studyView/StudyViewPageStore.ts [hunk 11]
```diff
@@ -229,47 +262,62 @@ export class StudyViewPageStore {
...
+            });
+        },
         default: [],
-        onResult:(attributes:ClinicalAttribute[])=>{
-            attributes.forEach(attribute=>{
-                this._attributeChartVisibility.set(attribute.clinicalAttributeId,false);
-            })
+        onResult:(result)=>{
+            _.forEach(result,obj=>{
+                this._clinicalAttributesMetaSet.set(obj.uniqueKey, obj)
```
[the remark this answers is not in the document]
Comment (kalletlak) [reply src/pages/studyView/StudyViewPageStore.ts]: there will be some actions, example changing chart type from pie to table, which is stored in the value via `chartType`. May be we can discuss this over chat.

File: src/pages/studyView/charts/ChartContainer.tsx [hunk 12]
```diff
@@ -91,49 +91,124 @@ export class ChartContainer extends React.Component<IChartContainerProps, {}> {
...
+                    active={this.mouseInChart}
+                    resetChart={this.handlers.resetFilters}
+                    deleteChart={this.handlers.onDeleteChart}
+                    hideLabel={this.hideLabel}
+                    chartControls={this.chartControls}
+                    changeChartType={this.changeChartType}
+                />
+
+                <If condition={this.props.chartMeta.chartType === ChartType.PIE_CHART}>
+                    <StudyViewComponentLoader promise={this.data}>
```
Comment (zhx828) [inline src/pages/studyView/charts/ChartContainer.tsx]: The if condition should be placed inside of StudyViewComponentLoader. This component will be used for all charts, then you can avoid duplication later.

File: src/pages/studyView/charts/StudyViewComponentLoader.tsx [hunk 13]
```diff
@@ -10,7 +10,7 @@ export interface StudyViewComponentLoaderProps {
 export class StudyViewComponentLoader extends React.Component<StudyViewComponentLoaderProps> {
     public render() {
         return (
-            <div style={{margin: 'auto'}}>
```
Comment (zhx828) [inline src/pages/studyView/charts/StudyViewComponentLoader.tsx]: Why did you remove the margin?

File: src/pages/studyView/charts/ChartContainer.tsx (hunk 12 above)
Comment (kalletlak) [reply src/pages/studyView/charts/ChartContainer.tsx]: if I do that and try changing the chart type, pie to table or vice-versa, nothing shows up.

File: src/pages/studyView/charts/ChartContainer.tsx [hunk 14]
```diff
@@ -1,38 +1,34 @@
 import * as React from "react";
```
Comment (zhx828) [inline src/pages/studyView/charts/ChartContainer.tsx]: This ChartContainer should soon be used by Scatter, Survival and Tables.

File: src/pages/studyView/StudyViewPage.tsx [hunk 15]
```diff
@@ -75,62 +76,57 @@ export default class StudyViewPage extends React.Component<IStudyViewPageProps,
...
-                                (
-                                    <div className={styles.studyViewFlexContainer}>
-                                        {this.store.visibleAttributes.map(this.renderAttributeChart)}
-                                    </div>
-                                )
-                            }
-                            {
-                                <div className={styles.studyViewFlexContainer}>
-                                    {this.store.survivalPlotData.result.map(this.renderSurvivalPlot)}
+        if (this.store.studies.isComplete) {
```
Comment (zhx828) [inline src/pages/studyView/StudyViewPage.tsx]: Should we add conditions when isPending and isError?

Comment (kalletlak) [reply src/pages/studyView/StudyViewPage.tsx]: I can add this in up coming pr, by showing a loading symbol while its still pending

File: src/pages/studyView/StudyViewPageStore.ts [hunk 16]
```diff
@@ -269,77 +262,71 @@ export class StudyViewPageStore {
...
+                    .get({ attribute: next, filters: this.emptyFilter })
                     .result!;
                 return acc;
             }, {} as any);
         },
         default: {},
-        onResult:() => {
-            this.initialized=true;
+        onResult: (result) => {
+            _.forEach(result, (obj, uniqueKey) => {
```
Comment (zhx828) [inline src/pages/studyView/StudyViewPageStore.ts]: Should we restart the conversation about the same uniqueKey(clinical attribute) in multiple studies?

Comment (kalletlak) [reply src/pages/studyView/StudyViewPageStore.ts]: Sure, I think its better to deal with this logic in backend

File: src/pages/studyView/StudyViewPageStore.ts [hunk 17]
```diff
@@ -269,77 +262,71 @@ export class StudyViewPageStore {
             let patientAttributeCount = 0;
             let filterAttributes: ClinicalAttribute[] = []
             queriedAttributes.forEach(attribute => {
-                if(attribute.patientAttribute){
-                    if(patientAttributeCount<10){
+                if (attribute.patientAttribute) {
+                    if (patientAttributeCount < 10) {
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts]: put comments explaining business logic (i.e. <10)

File: src/pages/studyView/StudyViewPageStore.ts [hunk 18]
```diff
@@ -269,77 +262,71 @@ export class StudyViewPageStore {
...
-        onResult:(attributes:ClinicalAttribute[])=>{
-            attributes.forEach(attribute=>{
-                this._attributeChartVisibility.set(attribute.clinicalAttributeId,true);
-            })
-        }
+        default: []
     });
 
-    readonly initalClinicalDataCounts = remoteData<{[id:string]:ClinicalDataCount[]}>({
+    readonly initalClinicalDataCounts = remoteData<{ [id: string]: ClinicalDataCount[] }>({
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts]: initial not "inital"

File: src/pages/studyView/StudyViewPageStore.ts [hunk 19]
```diff
@@ -179,84 +175,81 @@ export class StudyViewPageStore {
...
-    @computed get visibleAttributes(){
-        const attributes = this.clinicalAttributeSet.result
-        return _.reduce(this._attributeChartVisibility.toJS(),(acc:ClinicalAttribute[],next,key)=>{
-            if(next && attributes[key]){
-                acc.push(attributes[key])
-            }
-            return acc;
-        },[])
+    @computed get visibleAttributes(): ChartMeta[] {
+        if (this.initalClinicalDataCounts.isComplete) {
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts]: loading state should not handled inside of @computed getter.    this hides the loading state from calling code, which only knows whether array is empty or not (could be empty for other reasons).   Conditional should be in calling code.

**Comment by kalletlak** [comment]
@alisman Still there are lot of moving pieces. Will write tests once there is a stable code

File: src/pages/studyView/StudyViewPage.tsx [hunk 20]
```diff
@@ -75,62 +76,58 @@ export default class StudyViewPage extends React.Component<IStudyViewPageProps,
...
+                                <h3>Multiple Studies</h3>
+                            </If>
+                        </div>
+                    </div>
+                    <MSKTabs id="studyViewTabs" activeTabId={this.props.routing.location.query.tab}
+                        className="mainTabs">
+
+                        <MSKTab key={0} id="summaryTab" linkText="Summary">
+                            <div className={styles.studyViewFlexContainer}>
+                                {this.store.initialClinicalDataCounts.isComplete &&
```
Comment (alisman) [inline src/pages/studyView/StudyViewPage.tsx]: so, initialClinicalDataCounts will ensure the presenc eof visisbleAttributes, yes?  because onResult populates it?

Comment (kalletlak) [reply src/pages/studyView/StudyViewPage.tsx]: yes
